### PR TITLE
Part1

### PR DESCRIPTION
**Pull request by jasonp827**

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -109,18 +119,34 @@ int main()
...
+
+    for(Motorcycle& mc : motorcycles)
+    {
+        highway.addVehicle(&mc);
+    }
+
+    for(SemiTruck& st : trucks)
+    {
+        highway.addVehicle(&st);
+    }
```
Comment (matkatmusic) [inline main.cpp]: use `auto&` instead of specific types in case your container changes what it owns.
- [ ] revise

File: Vehicle.cpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    {
+        std::cout << "\nHighway is adjusting " << name << "'s speed to " << s << " mph" << std::endl;
+        speed = s;
+    }
+
+void Vehicle::tryToEvade()
+    {
+        std::cout << name << ": you'll never take me alive, ya dirty coppers!" << std::endl;
+        setSpeed(100);
+    }
```
Comment (matkatmusic) [inline Vehicle.cpp]: Fix this indentation
- [ ] revise

File: HighwayPatrol.cpp [hunk 3]
```diff
@@ -28,7 +31,7 @@ void HighwayPatrol::pullOver( Vehicle* v, bool willArrest, Highway* h )
     if( willArrest )
     {
         //print the vehicle type in this std::cout between "THE [" and "] PULL". 
-        std::cout << name << ": YOU IN THE [ " << " ] PULL OVER AND SHOW YOUR HANDS" << std::endl;
+        std::cout << name << ": YOU IN THE [ " << v->name <<  " ] PULL OVER AND SHOW YOUR HANDS" << std::endl;
```
Comment (matkatmusic) [inline HighwayPatrol.cpp]: `v->name` is not the `vehicle type`.   The purpose of this task in this function is for you to demonstrate knowledge of how to cast across the inheritance hierarchy to determine an object's type correctly.
- [ ] revise

File: main.cpp [hunk 4]
```diff
@@ -109,18 +119,34 @@ int main()
...
-    
-    
-    
-    
-    assert(false);
+
+     motorcycles.emplace_back("Ducati");
+     motorcycles.emplace_back("Harley Davidson");
+
+     trucks.emplace_back("SemiTruck");
```
Comment (matkatmusic) [inline main.cpp]: These need to be the names of drivers, not vehicle types.
- [ ] revise

File: SemiTruck.h [hunk 5]
```diff
@@ -1,2 +1,15 @@
...
+
+struct SemiTruck : public Vehicle
+{
+    SemiTruck(const std::string& n);
+    
+    void driveSafely();
+    void pullOver();
+
+private:
+    void tryToEvade() override;
```
Comment (matkatmusic) [inline SemiTruck.h]: Implement the special member functions as explained in `main.cpp`
remove `tryToEvade`
- [ ] revise

File: HighwayPatrol.h [hunk 6]
```diff
@@ -1,11 +1,14 @@
...
 
 #include "Vehicle.h"
+#include "Highway.h"
 
+struct Highway;
 struct HighwayPatrol : Vehicle 
 {
     HighwayPatrol();
+    ~HighwayPatrol() override;
     void scanHighway(Highway* h);
```
Comment (matkatmusic) [inline HighwayPatrol.h]: Implement the special member functions
- [ ] revise

File: Motorcycle.h [hunk 7]
```diff
@@ -1,7 +1,7 @@
 #pragma once
 
-#include <iostream>
 #include <limits>
+#include "Vehicle.h"
 
 struct Motorcycle : public Vehicle
 {
```
Comment (matkatmusic) [inline Motorcycle.h]: Implement the special member functions
- [ ] revise

File: Motorcycle.cpp [hunk 8]
```diff
@@ -17,11 +18,14 @@ void Motorcycle::setSpeed(int s)
 {
     if( s < 90 )
     {
+        Vehicle::setSpeed(110);
```
Comment (matkatmusic) [inline Motorcycle.cpp]: if `s < 90` this function ignores adjusting the speed to below 90.  you don't need this call to the base class function.
- [ ] remove it.

File: Highway.h [hunk 9]
```diff
@@ -2,15 +2,23 @@
 
 #include <vector>
 #include <algorithm>
+#include "Vehicle.h"
+#include "HighwayPatrol.h"
+
+struct Vehicle;
+struct HighwayPatrol;
```
Comment (matkatmusic) [inline Highway.h]: Don't forward declare and include.  pick one.   
You're not using either of these types in this header file, so **don't include what you don't use.**
- [ ] revise

File: Car.h [hunk 10]
```diff
@@ -1,9 +1,10 @@
...
 
 struct Car : public Vehicle
 {
-    Car(const std::string& s);
-    void closeWindows();
+    Car(const std::string& n);
 
-    void tryToEvade();
+    void closeWindows();
+    void tryToEvade() override;
```
Comment (matkatmusic) [inline Car.h]: implement the special member functions
- [ ] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: SemiTruck.h (hunk 5 above)
Comment (jasonp827) [reply SemiTruck.h:13]: Is it not ok to override tryToEvade and repurpose it for the pullOver func? Having a overridden destructor was producing some warnings

File: Car.h [hunk 11]
```diff
@@ -1,9 +1,14 @@
 #pragma once
+#include "Vehicle.h"
 
 struct Car : public Vehicle
```
Comment (matkatmusic) [inline Car.h]: struct is public by default.  this keyword is not needed
- [ ] revise

File: Motorcycle.h [hunk 12]
```diff
@@ -1,12 +1,16 @@
 #pragma once
 
-#include <iostream>
 #include <limits>
+#include "Vehicle.h"
 
 struct Motorcycle : public Vehicle
```
Comment (matkatmusic) [inline Motorcycle.h]: struct is public by default.  this keyword is not needed
- [ ] revise

File: SemiTruck.h [hunk 13]
```diff
@@ -1,2 +1,15 @@
 #pragma once
 
+#include "Vehicle.h"
+
+struct SemiTruck : public Vehicle
```
Comment (matkatmusic) [inline SemiTruck.h]: struct is public by default.  this keyword is not needed
- [ ] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**